<launch>
  <include file="$(find turtlebot_bringup)/launch/minimal.launch" />
  <include file="$(find turtlebot_teleop)/launch/myjoystick.launch" />
 
  <node pkg="tf" type="static_transform_publisher" name="base_to_laser" args="-0.13 0.0 0.28 0.0 0.0 0.0 base_link laser 100"/>   
 
  <arg name="map_file" default="/home/sam/map/sick_gmapping.yaml"/>
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />
  
 
  <!-- amcl reference -->
  <arg name="initial_pose_x"  default="0.0"/>
  <arg name="initial_pose_y"  default="0.0"/>
  <arg name="initial_pose_a"  default="0.0"/>
  <arg name="initial_cov_xx"  default="0.25"/>
  <arg name="initial_cov_yy"  default="0.25"/>
  <arg name="initial_cov_aa"  default="0.068538917"/>
 
  <arg name="use_map_topic"   default="true"/>
  <arg name="scan_topic"      default="scan"/> 
  <arg name="odom_topic" default="odom" />
  <arg name="laser_topic" default="scan" />
 
 
 
  <arg name="odom_frame_id"   default="odom"/>
  <arg name="base_frame_id"   default="base_footprint"/>
  <arg name="global_frame_id" default="map"/>

 
<node pkg="amcl" type="amcl" name="amcl" output="screen">
  <!-- Publish scans from best pose at a max of 10 Hz -->
  <param name="odom_model_type" value="diff"/>
  <param name="odom_alpha5" value="0.1"/>
  <param name="odom_alpha1" value="0.2"/>
  <param name="odom_alpha2" value="0.2"/>

  <param name="odom_alpha3" value="0.8"/>
  <param name="odom_alpha4" value="0.2"/>

  <param name="laser_max_beams" value="60"/>
  <param name="min_particles" value="1000"/>
  <param name="max_particles" value="10000"/>
  <param name="kld_err"       value="0.01"/>
  <param name="kld_z"         value="0.99"/>
  
  
  


  <!-- <param name="laser_z_hit" value="0.5"/>
  <param name="laser_z_short" value="0.05"/>
  <param name="laser_z_max" value="0.05"/>
  <param name="laser_z_rand" value="0.5"/> -->

  <param name="laser_sigma_hit" value="0.2"/>
 
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_model_type" value="likelihood_field"/>
  <!-- <param name="laser_model_type" value="beam"/> -->
  <param name="laser_likelihood_max_dist" value="2.0"/>
 
  
  <param name="resample_interval" value="3"/>
  <param name="transform_tolerance" value="0.15"/>
  <!-- <param name="recovery_alpha_slow" value="0.001"/>
  <param name="recovery_alpha_fast" value="0.1"/> -->
 
 
  <param name="first_map_only" value="true"/>
 
  <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
  <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
  <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
  <param name="initial_cov_xx"            value="$(arg initial_cov_xx)"/>
  <param name="initial_cov_yy"            value="$(arg initial_cov_yy)"/>
  <param name="initial_cov_aa"            value="$(arg initial_cov_aa)"/> 
 
  <param name="odom_frame_id"             value="$(arg odom_frame_id)"/> 
  <param name="base_frame_id"             value="$(arg base_frame_id)"/> 
  <param name="global_frame_id"           value="$(arg global_frame_id)"/>
  <param name="use_map_topic"             value="$(arg use_map_topic)"/>
  <remap from="scan"                      to="$(arg scan_topic)"/>
  
 
</node>



<include file="$(find charge_localization_and_move)/launch/includes/velocity_smoother.launch.xml"/>
<include file="$(find charge_localization_and_move)/launch/includes/safety_controller.launch.xml"/>

  <arg name="custom_param_file" default="$(find charge_localization_and_move)/param/dummy.yaml"/>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="$(find charge_localization_and_move)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find charge_localization_and_move)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />   
    <rosparam file="$(find charge_localization_and_move)/param/local_costmap_params.yaml" command="load" />   
    <rosparam file="$(find charge_localization_and_move)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find charge_localization_and_move)/param/dwa_local_planner_params.yaml" command="load" />
    <rosparam file="$(find charge_localization_and_move)/param/move_base_params.yaml" command="load" />
    <rosparam file="$(find charge_localization_and_move)/param/global_planner_params.yaml" command="load" />
    <rosparam file="$(find charge_localization_and_move)/param/navfn_global_planner_params.yaml" command="load" />
    <!-- external params file that could be loaded into the move_base namespace -->
    <rosparam file="$(arg custom_param_file)" command="load" />
    
    <!-- reset frame_id parameters using user input data -->
    <param name="global_costmap/global_frame" value="$(arg global_frame_id)"/>
    <param name="global_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="local_costmap/global_frame" value="$(arg odom_frame_id)"/>
    <param name="local_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="DWAPlannerROS/global_frame_id" value="$(arg odom_frame_id)"/>

    <remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <remap from="scan" to="$(arg laser_topic)"/>
  </node>

  <node pkg="charge_localization_and_move" type="odom2map" name="odom2map" output="screen" />
  <node pkg="charge_localization_and_move" type="showpose" name="showpose" output="screen" />
  <node pkg="charge_localization_and_move" type="showodom" name="showodom" output="screen" />
  <node pkg="charge_localization_and_move" type="showlaser" name="showlaser" output="screen" />
  <node pkg="charge_localization_and_move" type="showekf" name="showekf" output="screen" />
  
  <node pkg="charge_localization_and_move" type="charge_navigation_server" name="charge_navigation_server" output="screen" />
  <node pkg="charge_localization_and_move" type="charge_navigation_client" name="charge_navigation_client" output="screen" />
 
 

 <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf" output="screen" >
  <param name="output_frame" value="odom_combined"/>
  <param name="base_footprint_frame" value="base_footprint"/>
  <param name="freq" value="30.0"/>
  <param name="sensor_timeout" value="1.0"/>  

  <param name="odom_used" value="true"/>
  <param name="imu_used" value="true"/>
  <param name="vo_used" value="false"/>

  <remap from="imu_data" to="/mobile_base/sensors/imu_data" />


  </node>
 
</launch>